sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,792 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/rpcutil"
+
+	pbempty "github.com/golang/protobuf/ptypes/empty"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v2/proto/go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ProviderMain is the typical entrypoint for a resource provider plugin. Using it isn't required but can cut down
+// significantly on the amount of boilerplate necessary to fire up a new resource provider.
+func ProviderMain(name, version string, args ProviderArgs) error {
```

Comment:
I considered providing something that could work with this existing helper in pkg https://github.com/pulumi/pulumi/blob/master/pkg/resource/provider/main.go (possibly exposing a func from the Go sdk `pulumi` package that could be passed to `pkg/v2/resource/provider.Main`, but that felt a bit cumbersome, so went with a new "main" func here in `sdk` instead.

Reply:
I ended up reworking this so you can use it with the existing https://github.com/pulumi/pulumi/blob/master/pkg/resource/provider. I added a convenience `ComponentMain` there if you only care about a resource provider for components, but if you wanted to have a resource provider that had both custom and component resources that's now possible.